compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -2713,6 +2714,42 @@ private void checkConstantAssignment(BLangExpression varRef, AnalyzerData data)
         }
     }
 
+    private boolean isFiniteWithNilMember(BType bType) {
+        if (!(bType.getKind() == TypeKind.FINITE)) {
+            return false;
+        }
+        BFiniteType finiteType = (BFiniteType) bType;
+        Set<BLangExpression> finiteMembers = finiteType.getValueSpace();
+        for (BLangExpression finiteMemberType : finiteMembers) {
+            if (finiteMemberType.getBType() != null && finiteMemberType.getBType().tag != TypeTags.NIL) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean isUnionWithSingleNilTypeMember(BType bType) {
+        if (!(bType.getKind() == TypeKind.UNION)) {
+            return false;
+        }
+        BUnionType unionType = (BUnionType) bType;
+        LinkedHashSet<BType> members = unionType.getMemberTypes();
+        LinkedHashSet<BType> originalMembers = unionType.getOriginalMemberTypes();
+        if (members.size() >= 1) {
+            for (BType memberType : members) {
+                if (memberType.getKind() != TypeKind.NIL) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        // To bypass `function f3() returns never|never` kind of scenarios, referencing error: #39797
+        if (members.size() == 0 && originalMembers.size() == 1) {
```

Comment:
Why do we need `members.size() == 0` check? It is always true, right?

Reply:
Ideally, for a `union` type the  `unionType.getMemberTypes()` should not be empty. This needs to be fixed, so this check was to bypass the current status of the specific scenario.

Reply:
No need to create a special case for this scenario. If we modify the logic as I suggested, this case will handle correctly as well. Even after fixing this issue(https://github.com/ballerina-platform/ballerina-lang/issues/39797), there is no possibility of failure. (Please ensure that a test is added for this. Once the issue is resolved, we can verify if it has been fixed or not)
```java
 for (BType memberType : members) {
    if (memberType.tag != TypeTags.NIL) {
        return false;
    }
}
return true;
```

Reply:
The issue is how are we going to get the member type tag with `memberType.tag` by iterating `members` when `members` for cases where the return types is  like `never|never` is empty. In that case we need to iterate `getOriginalMemberTypes()`, that is why I cased it.